### DISCUSSION TITLE
Problem while installing PETSc libraries: PETSC_DIR already set

### DISCUSSION THREAD

**Discussion by hacs90** [fppimenta/rheoTool#62] [General]
Dear Prof F. Pinho,
sorry for bothering, but could you help me to solve an issue I had while trying to install rheoTool?. I tried to find solutions online  for my particular problem but I could not find any solid answer. 
I am following the user guide and trying to install rheoTool in my computer, but I had a problem while installing the PETSc libraries. After running the file "installPetsc", I received the following message in terminal: 
UNABLE to CONFIGURE with GIVEN OPTIONS (see configure.log for details):
--------------------------------------------------------------------------
The environmental variable PETSC_DIR /usr/lib/petscdir/3.7.7/x86_64-linux-gnu-real MUST be the current directory /home/hugo/OpenFOAM/hugo-8/ThirdParty/petsc-3.13.3.

I noticed that I already have a variable PETSC_DIR defined in my ~/.bashrc, and this is because I am currently using a CFD software called HiGFlow, which also uses the PETSc libraries (the software is using the version 3.7.7). The user guide recommends to be cautious while managing these variables because the file "installPetsc" might duplicate the PETSC_DIR variables. I am not an expert in bash scripting, so I don't know exactly how to do in this case but I tried to comment the variable #export PETSC_DIR = directory. However, this could not solve my issue. 

I would like to keep my previously set variable PETSC_DIR  that uses the version 3.7.7, and then define a new variable (let's say, PETSC_DIR2, which will be used by rheoTool as PETSC_DIR2 = /home/hugo/OpenFOAM/hugo-8/ThirdParty/petsc-3.13.3. Is this possible? I am thinking in modifying the file "installPetsc" and substitute PETSC_DIR2 for PETSC_DIR, but I am not sure if this is the only modification needed to be made there. Could this work? If that's not the case, could you please recommend me other alternatives?

I would highly appreciate your help. 
Best,
Hugo

**Comment by fppimenta** [comment]
Well, I don't know if you noticed my username, but it is not F Pinho. And BTW I am not professor...

Option 1:
Just try compiling rheoTool with your current version of petsc. It might work, although I am not sure. The oldest I have tried was 3.9.x if I remember well. That would be the easiest solution, as you would not have to do anything else, i.e. after downloading eigen, go directly to the _Allwmake_, skipping _installPetsc_.  But if your current petsc version doesn't have hypre bundled, then it will not work.

Option 2: 
To have more than 1 install of petsc you need a selector and there are several ways to do it. Assuming that PETSC_DIR and PETSC_ARCH for petsc 3.7.7 are defined in your .bashrc, then comment them with #. Then run installPetsc. If everything goes well you will end up with petsc 3.13 installed and your .bashrc will have:

#export PETSC_DIR=[something for 3.7.7]
#export PETSC_ARCH=[something for 3.7.7]

export PETSC_DIR=[something for 3.13]
export PETSC_ARCH=[something for 3.13]
export LD_LIBRARY_PATH=$PETSC_DIR/$PETSC_ARCH/lib:$LD_LIBRARY_PATH' 

Then install rheoTool with _Allwmake_. Now all you have to do is your selector, such that you can use 3.7.7 or 3.13 on demand. A simple way is to uncomment the lines for PETSC_DIR and PETSC_ARCH for 3.7.7 and put the new ones inside an alias, something like that:

alias of70='source /opt/openfoam7/etc/bashrc; export PETSC_DIR=[something for 3.13]; export PETSC_ARCH=[something for 3.13]; export LD_LIBRARY_PATH=$PETSC_DIR/$PETSC_ARCH/lib:$LD_LIBRARY_PATH' 

Then after opening a new terminal you should type _of70_ before using rheoTool, which will load the correct paths. You don't have to do anything to run your old programs, as the older petsc is 'loaded' by default.

If  PETSC_DIR and PETSC_ARCH for petsc 3.7.7 are not defined in your .bashrc, the rationale is the same: unset old vars -> set new vars, without forgetting LD_LIBRARY_PATH.

Option3:
If nothing else works, create a VM to run openfoam and rheoTool.

Extra: 
From your dir name it seems like you are using openfoam 8. rheoTool will not compile on that version. The most recent you can currently use is openfoam 7 (https://openfoam.org/version/7/).

**Reply by hacs90** [reply]
My apologies, I meant to write F P and not F Pinho.
Thank you for your suggestions and your help. I truly appreciate it. I tried option I and the PETSc 3.7.7 version did not work for rheoTool. I also tried option 2, I install the Eigen and PETSc libraries and create my alias, but when running the Allwmake file, it gives me an error indicating that probably the PETSc was not installed correctly (lnInclude/HyperIJSolver.H:43:10: fatal error: HYPRE_krylov.h: No such file or directory) even though my variables are well referenced: echo $PETSC_DIR = /home/hugo/openFOAM/hugo-7/ThirdPartt/petsc-3.13-3, echo $LD_LIBRARY_PATH , echo $PETSC_ARCH = arch-linux-c-opt. I tried to reinstall the libraries but it keeps giving me the same error. 

So I think the best option is trying to create a virtual machine.
Thanks again for all your advice.

**Comment by fppimenta** [comment]
Option 2 should work and I guess the error you are getting is specific of Hypre and not Petsc, i.e. if something was wrong with Petsc dirs and libs the error would be of another type. So my question is: are you sure that Hypre was successfully installed with Petsc? Are you able to run any Petsc tutorial/example using hypre? I would suggest you the following (you will just be trying a more recent version of petsc):
 - delete manually this folder /home/hugo/openFOAM/hugo-7/ThirdPartt/petsc-3.13-3
 - keep PETSC_DIR and PETSC_ARCH for 3.7.7 commented in .bashrc and delete all the lines related with 3.13.3, including the alias and the one related with LD_LIBRARY_PATH.
 - go to script _installPetsc_ and replace line 44 by: PetscV="3.15.1" 
 - run _installPetsc_ and in the phase it is downloading and installing third party soft like mumps, scotch, etc. be sure that **package Hypre is downloaded and installed correctly**. The information is displayed in your terminal. 
 - then run the Allwmake of rheoTool, but before that you should echo $PETSC_DIR,  $LD_LIBRARY_PATH and $PETSC_ARCH to verify that they point to the correct place.
 
If the above doesn't work, then send the configure.log created by Petsc and copy-paste the whole compilation result (error) of Allwmake.

**Reply by hacs90** [reply]
Hi. Thank you very much for your reply. You were right, I did not notice that there was a installation problem related to hypre. And I can confirm that by following your instructions, I finally was able to install rheoTool. When running Allwmake, it initially gave me an error telling me that it could not find the eigen files, but it was because I deleted the folder /home/hugo/openFOAM/hugo-7/ThirdPartt/petsc-3.13-3 and Eigen3.2.9 was there. But I reinstalled Eigen and run Allwmake and it compiled without problems. Now I am testing some tutorials examples and so far it is working fine.
Thanks again and sorry for taking some of your time.
Best,
Hugo
